### DISCUSSION TITLE
Is it possible to modify the type of property ?

### DISCUSSION THREAD

**Discussion by momala454** [CuyZ/Valinor#564] [Q&A] [answered]
I have a json that have a property that is (kind of) randomly an object instead of a string.
How can I detect when the property is an object and convert it to a string ? 
The object doesn't have a specific structure.

When I define the phpdoc of the property to `string|object`, it gives the error
```
   CuyZ\Valinor\Mapper\Tree\Exception\NoCasterForType 

  No caster was found to convert to type `object`.
```

And if I write it as `string`, obviously it will just says that it cannot be mapped ( CuyZ\Valinor\Mapper\TypeTreeMapperError)

Apparently, transformers are only used for serializer, and not for mapper, so I can't use it.
I found the `alter` method, but it will be called for every attribute, not just the one I want to modify.

Do you have any way to do that ?

**Comment by romm** [comment]
Hi @momala454, thanks for posting here.

A property typed as a `string|object` should accept any object value, as long as the `MapperBuilder` allows the permissive type `object`, using the `allowPermissiveTypes()` method. Could you post more information on what exactly the object coming from your "json" data is?

The `alter` method indeed exists and is part of the public API, although it is not documented (yet) as I am still not convinced I want it to stay it that way, so I would recommend not using it.

I may add attribute support for local alteration of a source value, but works on this task have not yet begun.

**Reply by momala454** [reply]
Hello @romm , thanks for your reply.
Here is a code to reproduce the issue.

```php
class Hello1
{
    /** @param array<string|object> $values */
    public function __construct(
        public array $values,
    ) {
    }
}
$json = '{"values": ["hello", {"world": "world"}]}';
$responseObject = (new \CuyZ\Valinor\MapperBuilder())
->allowSuperfluousKeys()
->allowPermissiveTypes()
->mapper()
->map(Hello1::class, \CuyZ\Valinor\Mapper\Source\Source::json($json));
print_r($responseObject);
```

edit: by the way, I would prefer to be able to only write `@param array<string>`, and only handle `object` values in a specific method to convert them to string before having any error

**Reply by romm** [reply]
Oh, now I get it.

When the JSON string is converted using the `\CuyZ\Valinor\Mapper\Source\Source::json` method, it becomes a nested array of scalars, there are no objects (note that this is due to [PHP's `json_decode` flag `associative`](https://www.php.net/en/json_decode)).

Therefore, what you need would be something like:

```php
class Hello1
{
    /** @param array<string|array> $values */
    public function __construct(
        public array $values,
    ) {}
}

$json = '{"values": ["hello", {"world": "world"}]}';
$responseObject = (new \CuyZ\Valinor\MapperBuilder())
    ->allowSuperfluousKeys()
    ->allowPermissiveTypes()
    ->mapper()
    ->map(Hello1::class, \CuyZ\Valinor\Mapper\Source\Source::json($json));
print_r($responseObject);
```

Also, using `allowPermissiveTypes` is not recommended as long as you don't need it. You may need it here, but if in your case you know the shape of the array, you could replace `array` by a shaped array type like `array{world: string}`, and remove the `allowPermissiveTypes` setting.

**Reply by momala454** [reply]
Ah, thank you !
And how do I modify the type of `$values` to be only string, and use something (normalizer?) to convert objects to string ? 
So when I use objects of the type `Hello1`, I know it will always be string ?
```
/** @param array<string> $values */
```

ps: the values when it's array is unknown, so I must use allowPermissiveType

**Reply by romm** [reply]
Arrays cannot be converted to strings by the mapper. I do not really see the point of what you're trying to achieve, could you be more specific about your real usecase?

**Reply by momala454** [reply]
of course by item you can't convert array to string, but I am looking for a way to specify a method to convert myself.
The use case is: an api return invalid values in some case (array instead of string), and I would like to "properly" convert to string to use it.
I suppose the normalizer feature is completely separated and cannot be used (like put an attribute with a custom class that allow me to convert array to string)

edit: i asked the api owner to fix it, and they told me it's not a priority to them. And I would like to have a clean class without "array" for "values" ;)

thanks :)

**Reply by romm** [reply]
I see.

The normalizer feature is indeed separated and cannot easily be used during the mapping process.

For your usecase, I'd probably do something like:

```php
class Hello1
{
    /** @var array<string> */
    public array $values;

    /** @param array<string|array> $values */
    public function __construct(
        array $values,
    ) {
        $this->values = array_map(fn ($value) => is_string($value) ? $value : json_encode($value), $values);
    }
}
```

**Reply by momala454** [reply]
thanks i will test it later.

**Reply by momala454** [reply]
@romm this solution works, but then it looks like the error reported is not precise when the type is nullable
```php
/** @property array<string> $values */
class Hello1
{
    /** @param array<string|array<string, mixed>> $values */
    public function __construct(
        public array $values,
        public ?Hello2 $world,
    ) {
        $this->values = array_map(fn ($value) => is_string($value) ? $value : json_encode($value, JSON_THROW_ON_ERROR), $values);
    }
}

class Hello2
{
    public function __construct(
        public string $value1,
        public string $value2,
        public string $value3,
    ) {
    }
}

$json = '{"values": ["hello", {"world": null}, {"hi": true}], "world": {"abc":"def"}}';
try {
$responseObject = (new \CuyZ\Valinor\MapperBuilder())
    ->allowSuperfluousKeys()
    ->allowPermissiveTypes()
    ->mapper()
    ->map(Hello1::class, \CuyZ\Valinor\Mapper\Source\Source::json($json));
    print_r($responseObject);

} catch (\CuyZ\Valinor\Mapper\TypeTreeMapperError $e) {
    echo 'messages:';
    $messages = \CuyZ\Valinor\Mapper\Tree\Message\Messages::flattenFromNode(
        $e->node()
    );
    var_dump($e->node()->name());
    foreach ($messages->errors() as $message) {
        var_dump($message->node()->path());
        var_dump($message->toString());
    }
    throw $e;
}
```

It simply says `Invalid value array{abc: 'def'}.`, but why is it invalid ? it doesn't say. I must change it to non nullable to see the real error : 
```
string(12) "world.value1"
string(71) "Cannot be empty and must be filled with a value matching type `string`."
string(12) "world.value2"
string(71) "Cannot be empty and must be filled with a value matching type `string`."
string(12) "world.value3"
string(71) "Cannot be empty and must be filled with a value matching type `string`."
```

If I must create an issue, please tell me

**Reply by romm** [reply]
The signature of `Hello2` is invalid. The PHPDoc indicates a type for a non-existing parameter `$values`. Something like below would work:

```php
class Hello2
{
    /** @param array<string|array<string, mixed>> $values */
    public function __construct(
        public array $values,
    ) {}
}
```

**Reply by momala454** [reply]
sorry that was a copy/paste error. Remove the phpdoc for Hello2 (i've updated my code above). The problem I mentioned stays the same

**Comment by romm** [comment, marked as answer]
Closing, as converters should now solve this particular usecase, see [documentation about this](https://github.com/CuyZ/Valinor/blob/0a8c0d0574690ebf70e3959a2438dcf83ce58e5c/docs/pages/how-to/common-converters-examples.md) for more info.

**Reply by momala454** [reply]
great thanks
